### PR TITLE
AZ-663 Added support for bond, set_keys, rotate_keys, validate, transfer, set_staking_limits, force_new_era commands in cliain

### PR DESCRIPTION
**Pull request by Marcin-Radecki**

### CODE REVIEW

File: bin/cliain/src/validators.rs [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
-    ).collect();
+    let validators: Vec<_> = validators
+        .into_iter()
+        .map(|validator| AccountId::from_ss58check(&validator).expect("Address is valid"))
+        .collect();
 
     change_members(&connection, validators, XtStatus::Finalized);
-
-    info!("Validators changed")
+    // TODO we need to check state here whether change members actually succeed
```
Comment (mike1729) [inline bin/cliain/src/validators.rs:15]: Better add jira ticket than a TODO.

**Review by mike1729 — COMMENTED** [review]
Only nits. 

One major remark is that expect messages are very inconsistent, some declare what went wrong and some why it won't panic. This inconsistency is however very consistent with the whole repo.

File: bin/cliain/src/main.rs [hunk 2]
```diff
@@ -23,28 +28,135 @@ struct Config {
...
         validators: Vec<String>,
     },
+
+    /// Force new era in staking world. Requires sudo.
+    ForceNewEra,
+
     /// Associate the node with a specific staking account.
     PrepareKeys,
+
+    /// call rotate_keys() RPC call and prints them to stdout
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply bin/cliain/src/main.rs:59]: Fixed

File: bin/cliain/src/main.rs [hunk 3]
```diff
@@ -23,28 +28,135 @@ struct Config {
...
+    Bond {
+        /// Seed for a stash account
+        #[clap(long)]
+        stash_seed: String,
+
+        /// SS58 id of the controller account
+        #[clap(long)]
+        controller_account: String,
+
+        /// a stake to bond (in tokens)
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply bin/cliain/src/main.rs:41]: Fixed

File: bin/cliain/src/main.rs [hunk 4]
```diff
@@ -23,28 +28,135 @@ struct Config {
...
+
     /// Associate the node with a specific staking account.
     PrepareKeys,
+
+    /// call rotate_keys() RPC call and prints them to stdout
+    RotateKeys,
+
+    /// Sets given keys for origin controller
+    SetKeys {
+        /// a 64 byte hex encoded string in form 0xaabbcc..
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply bin/cliain/src/main.rs:64]: Fixed

File: bin/cliain/src/main.rs [hunk 5]
```diff
@@ -23,28 +28,135 @@ struct Config {
...
+        /// where aabbcc...  must be exactly 128 characters long
+        #[clap(long)]
+        new_keys: String,
+
+        /// seed for a controller account which signes set_keys tx
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply bin/cliain/src/main.rs:69]: Fixed

File: bin/cliain/src/main.rs [hunk 6]
```diff
@@ -23,28 +28,135 @@ struct Config {
...
+        /// Number of tokens to send,
+        #[clap(long)]
+        amount_in_tokens: u64,
+
+        /// SS58 id of target account
+        #[clap(long)]
+        to_account: String,
+    },
+
+    /// call staking validate call for a given controller
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply bin/cliain/src/main.rs:100]: Fixed

File: bin/cliain/src/main.rs [hunk 7]
```diff
@@ -23,28 +28,135 @@ struct Config {
...
+    Validate {
+        /// seed for a controller account to intent being validator to
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply bin/cliain/src/main.rs:102]: Fixed

File: bin/cliain/src/validators.rs (hunk 1 above)
Comment (Marcin-Radecki) [reply bin/cliain/src/validators.rs:15]: Good point, raised https://cardinal-cryptography.atlassian.net/browse/AZ-699 and updated comment.

File: bin/cliain/src/lib.rs [hunk 8]
```diff
@@ -1,5 +1,16 @@
...
 
-pub use validators::change as change_validators;
 pub use keys::prepare as prepare_keys;
+pub use keys::rotate_keys_command as rotate_keys;
+pub use keys::set_keys_command as set_keys;
+pub use secret::prompt_password_hidden;
+pub use staking::bond_command as bond;
+pub use staking::force_new_era_command as force_new_era;
+pub use staking::set_staking_limits_command as set_staking_limits;
+pub use staking::validate_command as validate;
```
Comment (timorl) [inline bin/cliain/src/lib.rs:14]: Collapse, although see my earlier comment about good reasons.

File: bin/cliain/src/main.rs [hunk 9]
```diff
@@ -13,38 +17,175 @@ struct Config {
...
+        /// Seed for a controller account which signes set_keys tx
+        #[clap(long)]
+        controller_seed: String,
+    },
+
+    /// Command to convert given seed to SS58 Account id
+    SeedToSS58 {
+        /// Seed of the account to convert to SS58 public key to
+        #[clap(long)]
+        seed: String,
```
Comment (timorl) [inline bin/cliain/src/main.rs:78]: I'm pretty sure you can just use the global `--seed` flag and then there should be no problem with treating this as a normal subcommand?

**Review by timorl — CHANGES_REQUESTED** [review]
Noice.

File: aleph-client/src/lib.rs [hunk 10]
```diff
@@ -7,14 +7,21 @@ use substrate_api_client::{rpc::ws_client::WsRpcClient, Api, RpcClient, XtStatus
 mod rpc;
 mod session;
 mod staking;
+mod transfer;
 mod waiting;
 
 pub use rpc::rotate_keys;
+pub use rpc::rotate_keys_raw;
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply aleph-client/src/lib.rs:14]: Done

File: aleph-client/src/lib.rs [hunk 11]
```diff
@@ -7,14 +7,21 @@ use substrate_api_client::{rpc::ws_client::WsRpcClient, Api, RpcClient, XtStatus
...
 pub use session::{
     change_members, get_current as get_current_session, set_keys, wait_for as wait_for_session,
     Keys as SessionKeys,
 };
 pub use staking::bond as staking_bond;
+pub use staking::force_new_era as staking_force_new_era;
+pub use staking::set_staking_limit as staking_set_staking_limits;
+pub use staking::validate as staking_validate;
+pub use transfer::transfer as balances_transfer;
+pub use transfer::TransferTransaction;
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply aleph-client/src/lib.rs:24]: Good point, done.

File: aleph-client/src/rpc.rs [hunk 12]
```diff
@@ -16,11 +16,13 @@ pub fn author_rotate_keys() -> Value {
...
-    }))
+    Ok(connection
+        .get_request(author_rotate_keys())?
+        .map(|keys| SessionKeys::from(keys)))
+}
+
+pub fn rotate_keys_raw(connection: &Connection) -> ApiResult<Option<String>> {
+    Ok(connection
+        .get_request(author_rotate_keys())?
+        .map(|keys| keys.trim_matches('\"').to_string()))
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply aleph-client/src/rpc.rs:27]: I extracted a common method to remove duplication. Also, I've added comment to explain we need to escape some characters from RPC result.

File: aleph-client/src/session.rs [hunk 13]
```diff
@@ -23,6 +24,14 @@ impl From<Vec<u8>> for Keys {
     }
 }
 
+impl From<String> for Keys {
+    fn from(keys: String) -> Self {
+        let bytes: Vec<u8> =
+            FromHexString::from_hex(keys).expect("String hex-encoded session cannot be decoded");
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply aleph-client/src/session.rs:30]: Good catch, we don't want `expect` here as we don't want to crash an app in such a scenario. Change to `TryFrom`.

File: aleph-client/src/session.rs [hunk 14]
```diff
@@ -1,8 +1,9 @@
-use crate::{send_xt, wait_for_event, BlockNumber, Connection};
+use crate::waiting::wait_for_event;
+use crate::{send_xt, BlockNumber, Connection};
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply aleph-client/src/session.rs:2]: Done.

File: benches/payout-stakers/src/main.rs [hunk 15]
```diff
@@ -1,23 +1,22 @@
...
 use rayon::prelude::*;
-use sp_core::sr25519::Pair;
+use sp_core::sr25519::Pair as KeyPair;
+use sp_core::Pair;
 use sp_keyring::AccountKeyring;
 use std::iter;
-use substrate_api_client::{rpc::WsRpcClient, Api, XtStatus};
-use aleph_client::staking_bond;
+use substrate_api_client::extrinsic::staking::RewardDestination;
+use substrate_api_client::{rpc::WsRpcClient, AccountId, Api, XtStatus};
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply benches/payout-stakers/src/main.rs:19]: Done

File: bin/cliain/README.md [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+
+## Overview
+
+`cliain` is a wrapper over `substrate-api-client` library. It makes easy to call some Substrate chain
+extrinsic or RPC calls. Run `./cliain --help` to see which of them are supported.
+
+## Signing account
+
+Tool reqires `--seed` parameter to sign given transaction with an account derived from the given seed.
+If `--seed` is not given, root account is assumed and prompt is displayed to enter root seed.
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply bin/cliain/README.md:15]: Good point, changed

File: bin/cliain/src/lib.rs (hunk 8 above)
Comment (Marcin-Radecki) [reply bin/cliain/src/lib.rs:14]: Done.

File: bin/cliain/src/main.rs (hunk 9 above)
Comment (Marcin-Radecki) [reply bin/cliain/src/main.rs:78]: Right, I was blind, changed.

**Comment by Marcin-Radecki** [comment]
Thanks for the review, the code is better now.

**Review by mike1729 — APPROVED** [review]
g2g

**Pull request merged**